org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModelManager.java

```diff
@@ -4831,22 +4902,20 @@ public void secondaryTypeAdding(String path, char[] typeName, char[] packageName
 	public Map<String, Map<String, IType>> secondaryTypes(IJavaProject project, boolean waitForIndexes, IProgressMonitor monitor) throws JavaModelException {
 		if (VERBOSE) {
 			StringBuilder buffer = new StringBuilder("JavaModelManager.secondaryTypes("); //$NON-NLS-1$
-			buffer.append(project.getElementName());
-			buffer.append(',');
-			buffer.append(waitForIndexes);
-			buffer.append(')');
+			buffer.append(project.getElementName()).append(',').append(waitForIndexes).append(')');
 			Util.verbose(buffer.toString());
 		}
 
 		// Return cache if not empty and there's no new secondary types created during indexing
 		final PerProjectInfo projectInfo = getPerProjectInfoCheckExistence(project.getProject());
-		Map<IFile, Map<String, Map<String, IType>>> indexingSecondaryCache = projectInfo.secondaryTypes == null ? null : projectInfo.indexingSecondaryCache;
-		if (projectInfo.secondaryTypes != null && indexingSecondaryCache == null) {
-			return projectInfo.secondaryTypes;
+		SecondaryTypesCache secondaryTypesCache = projectInfo.secondaryTypes.cache();
```

Comment:
i assume this and the other 3 calls can not happen without getOrCreate be called before?

Reply:
which other 3 calls do you mean?
The NP check happens in the `if(secondaryTypesCache.isIndexingDone())` and below.
So if the secondaryTypes map isn't created yet, we will start search.

Reply:
no i mean cache() can never return null